### ISSUE TITLE
questions/improvements for tests/run.sh

### ISSUE THREAD

**Issue by benaryorg** [r0x0r/pywebview#1525]
### Specification
- pywebview version: master
- operating system: unixoid
- web renderer: any

### Description

I'm currently checking out a pywebview build error on NixOS and I'm seeing several workarounds for things done in *tests/run.sh* and I'm also seeing the semi-recent commits to improve the tests and I just wanted to ask what the purpose of that file is and whether I can help improve it.

As far as I can see the entire *run.sh* can be reduced down to:

```bash
#!/usr/bin/env bash

export PYWEBVIEW_LOG='debug'
python3 -m pytest -q -s --disable-warnings -r w "$@"
```

However that would change the model of running the tests to running the entire test suite once instead of invoking the tests individually, and historically the file has used various ways to first collect the list of tests (earlier using *pytest* itself, now using a (needless) subshell) and I really wanted to know whether there is a reason for doing that?

There are several issues with the current *run.sh*;

1. [environment variables aren't exported](https://github.com/r0x0r/pywebview/blob/ba0880cf74334af6ec21670bec8cea6de4cdff5b/tests/run.sh#L3) which should mean that setting that shell variable effectively becomes a no-op for that suite, which I assume is not the intention
1. [the file listing is subject to string splitting](https://github.com/r0x0r/pywebview/blob/ba0880cf74334af6ec21670bec8cea6de4cdff5b/tests/run.sh#L16), so if any test file ever contains for instance a space, this will break. An easy fix here is to just use `for test in test_*; do`, but again, if you intend to always run all tests why not use pytest on the entire project anyway? (similarly [the use of the variable should be quoted](https://github.com/r0x0r/pywebview/blob/ba0880cf74334af6ec21670bec8cea6de4cdff5b/tests/run.sh#L18) as `"$test"`)
1. `#!/usr/bin/env bash` is the correct shebang for *bash*

And then as a package maintainer it'd be cool if one could pass in additional *pytest* settings too, which is what the `"$@"` in the proposed replacement above does. That way packagers can add options like `--ignore=some_failing_test` so that tests can still be run on the distro side even when individual tests are known to fail at time of packaging, or if the distro needs to pass some other options for whatever reason.

Basically, what I'm saying is, what are you trying to accomplish with the *run.sh* file, so that I can potentially send a PR that declutters it a bit and scratches an itch from the maintainer side.?

### Practicalities

- YES I am willing to work on this issue myself.

- NO I am prepared to support this issue financially.

**Comment by r0x0r** [comment]
The motivation behind `run.sh` was to run each test in a separate process. At least previousls tests tended to stuck when run under pytest in batch. I have recently made some improvements to cleaning the state of pywebview after closing the main window, but I have not really investigated whether `run.sh` is still needed.

As for PR, I am all up for streamlining tests, but it should be verified that tests can be run for all the platforms both locally and under the CI/CD environment as well.

**Comment by r0x0r** [comment]
I got rid of run.sh / run.ps1 in this PR  https://github.com/r0x0r/pywebview/pull/1533 along with the improved test infrastructure and event handling

**Comment by benaryorg** [comment]
Aight, I was still keeping this in the back of my head but didn't get around to actually doing much, so I guess you beat me to it.

Since that PR gets rid of the *run.sh* entirely and replaces it with a blanket call to pytest I guess downstream nixpkgs can also just do that (with the boilerplate of using xvfb-run and pulling in Qt and all that) to mask any tests we know to be failing for whatever reasons (Nix being special sometimes) then.
Ideally we would just use the regular testing infrastructure but the Qt wrapping there won't disappear any time soon, same for the xvfb-run, so this is as good as it gets there.
Basically calling *pytest* directly allows passing `--deselect` to deselect individual tests which was all we really need.

Looking at the changes I see that the powershell version still calls the tests individually.
Do things break on Windows when calling the tests via one big pytest invocation?
I don't have any Windows machine anymore, so I couldn't test there without setting up a full VM or something, which is why I didn't really get around to testing a few things in that respect.
